pkg/registry/core/service/strategy.go

```diff
@@ -205,6 +207,127 @@ func topologyKeysInUse(svc *api.Service) bool {
 	return len(svc.Spec.TopologyKeys) > 0
 }
 
+// This is an unusual case.  Service has a number of inter-related fields and
+// in order to avoid breaking clients we try really hard to infer what users
+// mean when they change them.
+//
+// Services are effectively a discriminated union, where `type` is the
+// discriminator. Some fields just don't make sense with some types, so we
+// clear them.
+//
+// As a rule, we almost never change user input.  This can get tricky when APIs
+// evolve and new dependent fields are added.  This specific case includes
+// fields that are allocated from a pool and need to be released.  Anyone who
+// is contemplating copying this pattern should think REALLY hard about almost
+// any other option.
+func dropTypeDependentFields(newSvc *api.Service, oldSvc *api.Service) {
+	// For now we are only normalizing on updates.  This minimizes potential
+	// confusion since many of the cases we are handling here are pretty niche.
+	if oldSvc == nil {
+		return
+	}
+
+	// If the user is switching to a type that does not need a value (even
+	// "None" counts), AND they did not change the ClusterIP value, we can wipe
+	// it.  It will be deallocated later.
+	//
+	// There are other fields that could also be wiped, but historically were
+	// ignored.  For compat, we leave it that way.
+	if needsClusterIP(oldSvc) && !needsClusterIP(newSvc) && sameClusterIP(oldSvc, newSvc) {
+		newSvc.Spec.ClusterIP = ""
+	}
+
+	// If the user is switching to a type that doesn't use NodePorts AND they
+	// did not change any NodePort values, we can wipe them.  It will be
+	// deallocated later.
+	if needsNodePort(oldSvc) && !needsNodePort(newSvc) && sameNodePorts(oldSvc, newSvc) {
+		for i := range newSvc.Spec.Ports {
+			newSvc.Spec.Ports[i].NodePort = 0
+		}
+	}
+
+	// If the user is switching to a case that doesn't use HealthCheckNodePort AND they
+	// did not change the HealthCheckNodePort value, we can wipe it.  It will
+	// be deallocated later.
+	if needsHCNodePort(oldSvc) && !needsHCNodePort(newSvc) && sameHCNodePort(oldSvc, newSvc) {
+		newSvc.Spec.HealthCheckNodePort = 0
+	}
+
+	// NOTE: there are other fields like `selector` which we could wipe.
+	// Historically we did not wipe them and they are not allocated from
+	// finite pools, so we are (currently) choosing to leave them alone.
+}
+
+func needsClusterIP(svc *api.Service) bool {
+	if svc.Spec.Type == api.ServiceTypeExternalName {
+		return false
+	}
+	return true
+}
+
+func sameClusterIP(oldSvc, newSvc *api.Service) bool {
+	return oldSvc.Spec.ClusterIP == newSvc.Spec.ClusterIP
+}
+
+func needsNodePort(svc *api.Service) bool {
+	if svc.Spec.Type == api.ServiceTypeNodePort || svc.Spec.Type == api.ServiceTypeLoadBalancer {
+		return true
+	}
+	return false
+}
+
+func sameNodePorts(oldSvc, newSvc *api.Service) bool {
+	// Helper to make a unique key so we can tell if a port changed.
+	makeKey := func(p *api.ServicePort) string {
+		return fmt.Sprintf("%s/%d", p.Protocol, p.Port)
```

Comment:
I thought about that and it seems correct.  It is effectively a new port definition.  So if you did this AND switch to ExternalName at the same time, it would fail (later in validation) for having a NodePort value.

We simply don't have an immutable key for ports, so I can't really tell the difference between "I changed an existing port number which already had a NodePort" and "I added a new port and specified a NodePort for it".  I think this is as close to correct as we can get.

There's a test-case for it "don't clear changed nodePorts" and "don't clear nodePorts when adding a port with NodePort"

Reply:
seems reasonable, just wanted to make sure we considered it